flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -623,37 +720,60 @@ class ExprLowering {
   }
 
   template <typename A>
-  mlir::Value genval(const Fortran::evaluate::ArrayConstructor<A> &) {
+  Fortran::lower::ExValue
+  genval(const Fortran::evaluate::ArrayConstructor<A> &) {
     TODO();
   }
-  mlir::Value gen(const Fortran::evaluate::ComplexPart &) { TODO(); }
-  mlir::Value gendef(const Fortran::evaluate::ComplexPart &cp) {
-    return gen(cp);
+
+  Fortran::lower::ExValue gen(const Fortran::evaluate::ComplexPart &) {
+    TODO();
+  }
+  Fortran::lower::ExValue genval(const Fortran::evaluate::ComplexPart &) {
+    TODO();
   }
-  mlir::Value genval(const Fortran::evaluate::ComplexPart &) { TODO(); }
 
-  mlir::Value gen(const Fortran::evaluate::Substring &s) {
+  /// Reference to a substring.
+  Fortran::lower::ExValue gen(const Fortran::evaluate::Substring &s) {
     // Get base string
     auto baseString = std::visit(
         Fortran::common::visitors{
             [&](const Fortran::evaluate::DataRef &x) { return gen(x); },
             [&](const Fortran::evaluate::StaticDataObject::Pointer &)
-                -> mlir::Value { TODO(); },
+                -> Fortran::lower::ExValue { TODO(); },
         },
         s.parent());
     llvm::SmallVector<mlir::Value, 2> bounds;
-    bounds.push_back(genval(s.lower()));
-    if (auto upperBound{s.upper()}) {
-      bounds.push_back(genval(*upperBound));
+    auto lower = genunbox(s.lower());
+    assert(lower && "boxed value not handled");
+    bounds.push_back(lower);
+    if (auto upperBound = s.upper()) {
+      auto upper = genunbox(*upperBound);
+      assert(upper && "boxed value not handled");
+      bounds.push_back(upper);
     }
-    return builder.createSubstring(baseString, bounds);
+    // FIXME: a string should be a CharBoxValue
+    auto addr = fir::getBase(baseString);
+    return builder.createSubstring(addr, bounds);
   }
 
-  mlir::Value gendef(const Fortran::evaluate::Substring &ss) { return gen(ss); }
-  mlir::Value genval(const Fortran::evaluate::Substring &ss) { return gen(ss); }
-  mlir::Value genval(const Fortran::evaluate::Triplet &trip) { TODO(); }
+  /// The value of a substring.
+  Fortran::lower::ExValue genval(const Fortran::evaluate::Substring &ss) {
+    // FIXME: why is the value of a substring being lowered the same as the
+    // address of a substring?
+    return gen(ss);
+  }
+
+  Fortran::lower::ExValue genval(const Fortran::evaluate::Triplet &trip) {
```

Comment:
A Triplet is not an expression on its own contrary to many other `Fortran::evaluate::XX`, as such I think it would be fine if it returned something other than a  `Fortran::lower::ExValue`. It would allow not putting the `RangeBoxValue` under `Fortran::lower::ExValue` which I find a bit weird given all the other variant of the union represents "full" expressions, but this one is at the same level but only contains a piece of an expressions that does not make sens on its own.

Maybe the function should be call `genTriplet` instead of genval to differentiate it and it could return a RangeBoxValue.
I understand this is easier said than done given this function is accessed through std::visit, but , maybe `genval(const Fortran::evaluate::Subscript &subs)` could also be modified here to not return an `Fortran::lower::ExValue` but a `SubscriptValue` that would be a union between `RangeBoxValue`, `UnBoxValue` and something for vector subscript and that could be defined and used locally instead of using ExValue to propagate this.

Reply:
I tried to answer that here. https://github.com/flang-compiler/f18-llvm-project/pull/75#discussion_r427398802

It may be that the adding triples as special cases would improve something in the code. On the other hand, passing a triple as another case of a boxed (aggregate) value seems fine to me.